core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -854,7 +854,8 @@ private[spark] object Utils extends Logging {
    * Get the local machine's URI.
    */
   def localHostNameForURI(): String = {
-    customHostname.getOrElse(InetAddresses.toUriString(localIpAddress))
+    sys.env.get("SPARK_PUBLIC_DNS").getOrElse(
```

Comment:
@tnachen writing from what I remember the issue here was, I can retest if necessary but it's a rather long process:

`customHostname` comes from `SPARK_LOCAL_HOSTNAME `. When I was implementing this code, I tried setting the value of `SPARK_LOCAL_HOSTNAME ` to the same value as `SPARK_PUBLIC_DNS`. However, after getting to the `localHostNameForURI()` call, the `customHostname` value was no longer what I set `SPARK_LOCAL_HOSTNAME`. In other words, it has been changed by something from outside and I am not able to find it what happened.

Let me ask a question: under what circumstances the user would like to set `SPARK_LOCAL_HOSTNAME` to a different value than `SPARK_PUBLIC_DNS `?